tests/internal/utils/utils.go

```diff
@@ -0,0 +1,124 @@
+package utils
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+	"testing"
+	"time"
+)
+
+// T wraps testing.T and the configurations of the testing instance.
+type T struct {
+	*testing.T
+	Config *Config
+}
+
+// New create an instance of T
+func New(t *testing.T, c *Config) *T {
+	return &T{T: t, Config: c}
+}
+
+// Config Settings of the testing program
+type Config struct {
+	// The executable path of the tested program.
+	Program string
+	// Working directory prepared for the tested program.
+	// If empty, a directory named with random suffixes is picked, and created under the current directory.
+	// The directory will be removed when the test finishes.
+	WorkDir string
+	// Command-line arguments passed to the tested program.
+	Args []string
+
+	// Where to redirect the stdout/stderr to. For debugging purposes.
+	LogFile *os.File
+}
+
+func redirect(cmd *exec.Cmd, f *os.File) error {
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+
+	go io.Copy(os.Stderr, stdout)
+	go io.Copy(os.Stdout, stderr)
+	return nil
+}
+
+// RunTest Helper function for setting up a running Gitea server for functional testing and then gracefully terminating it.
+func (t *T) RunTest(tests ...func(*T) error) (err error) {
+	if t.Config.Program == "" {
+		return errors.New("Need input file")
+	}
+
+	path, err := filepath.Abs(t.Config.Program)
+	if err != nil {
+		return err
+	}
+
+	workdir := t.Config.WorkDir
+	if workdir == "" {
+		workdir, err = filepath.Abs(fmt.Sprintf("%s-%10d", filepath.Base(t.Config.Program), time.Now().UnixNano()))
```

Comment:
I'd prefer something like `ioutil.TempDir(os.TempDir(), "gitea_tests")`. Since we're going to discard the contents of `workdir` anyways (line 78), we might as well use a temporary directory.